### ISSUE TITLE
Composite types can not be safely encode or decoded

### ISSUE THREAD

**Issue by fishcakez** [elixir-ecto/postgrex#353]
If a composite type, such as a table row, changes its types then we do not update the types and oids. For encoding this means we send the wrong types to the database and error. For decoding it means we crash because oids don't match. Similarly for decoding anonymous composites it possible that it includes a type that we have not bootstrapped because the preparation of the query won't signal the missing type. In this situation we crash too. Catching any of these situations in a test is very unlikely.

I am unsure if this has been reported(https://github.com/elixir-ecto/ecto/issues/1271#issuecomment-343716476) . It would be very difficult to diagnose because of the complicated way postgrex handles types. Also I am unsure how much composites are used.

For encoding we can resolve this by looking up the composite types at parameter describe time - we don't support anonymous composites. If we require a redescribe (executing on new connection with known parameters types but unknown result types) and types changed then we will get a postgresql error because we don't lookup parameters on redescribe (1eb868). Doing a query on every composite prepare is undesirable but should work. We would be logging a warning every time this occurred. For performance people may choose not to use this but existing statements would still work.

For decoding we can partially resolve the internal types changing by treating all composites types as anonymous. This means we delay resolving the oids until decoding, or perhaps only if oids have changed fallback to the anonymous approach. However this means we need to be able to decode anonymous types.

Alternatively, or additional, we could lookup the oids at describe time (we describe results on explicit prepare and implicit redescribe). Unfortunately postgresql doesn't invalidated prepared queries when the sub-types change so we would either need to fallback to anonymous or crash if types change.

For decoding anonymous composites I am not sure. Technically if we find a type we don't know during decode we could look it up after results finish and then decode again. That'd be complex.

I feel this issue invalidates our general type handling approach somewhat and means edge cases will always exist. We may wish to consider having an additional type module that behaves like the majority of drivers in other languages: binary for built in types, text for other types. Of course we are making a trade off to get a richer type system so it might not be worthwhile changing existing behavior.

**Comment by michalmuskala** [comment]
I wonder if it maybe would make sense to have some explicit API to tell postgrex that types have changed. I can see this being used when you migrate your database and then call some API in ecto to tell it about this. This could potentially give us some way of handling this - especially if we're saying we can never avoid all the edge cases.

**Comment by josevalim** [comment]
@fishcakez is this now easier with DBConnection.EncodeError?

**Comment by edgurgel** [comment]
Hey team I'm here just to say we got hit by this bug. We ended up not relying on Postgrex and just sending raw data over the wire: https://github.com/supabase/realtime/pull/1981/changes#diff-b4b8bf1476bfaee70b2992b7829540b88a87bbd9cec8573837dc55ef7cbee482R83-R98

**Comment by Schultzer** [comment]
> Hey team I'm here just to say we got hit by this bug. We ended up not relying on Postgrex and just sending raw data over the wire: https://github.com/supabase/realtime/pull/1981/changes#diff-b4b8bf1476bfaee70b2992b7829540b88a87bbd9cec8573837dc55ef7cbee482R83-R98

Do I understand this correctly, that this happens when you don’t control migrations, as in a typical Ecto app you would do a new release to apply migrations and then the types would be updated.

**Comment by edgurgel** [comment]
> > Hey team I'm here just to say we got hit by this bug. We ended up not relying on Postgrex and just sending raw data over the wire: https://github.com/supabase/realtime/pull/1981/changes#diff-b4b8bf1476bfaee70b2992b7829540b88a87bbd9cec8573837dc55ef7cbee482R83-R98
> 
> Do I understand this correctly, that this happens when you don’t control migrations, as in a typical Ecto app you would do a new release to apply migrations and then the types would be updated.

Sort of! The issue we found was that we used the old code with old type, kept the connection up, ran the migrations, then ran the new code with the new type and the query fails. It only succeeds after killing the connection pool if I am not mistaken.

So it wont self-heal without intervention.

So you need to have a clear deploy that recycles all connections before you use the updated composite type

**Comment by josevalim** [comment]
Correct. I believe you can use disconnect_all to trigger the reaping too.

**Comment by greg-rychlewski** [comment]
Would there be a race condition with disconnect all where some connections might be restarted before all of them are gone?

Some drivers like pgx in go have a type loader function that you can call from your migration to create/refresh the type cache for specific types. This would be similar to what Michal suggested. I could implement that if it would be useful

**Comment by Schultzer** [comment]
Hmm regardless of the types being tied to the connection, then there is still the issue of running migrations with new code and still have old nodes references old types or columns in a cluster.

I believe that in theory migrations could be run with global in a distributed Elixir cluster in a combination of hot code upgrade, the pool would have to make callers queue up until done. Although that doesn’t protect you from long running migrations.
